New Flax release: v0.3.4 (#1318)
google/flax: CHANGELOG.md

(Add your change to a random empty line to avoid merge conflicts)
 -
 -
 -
 -
 -
 -
 -
 -
 -
 -
 -
 -
 -
 -
 -
 -
 -
 -
 -
 -
 -
 -
 -

0.3.4
------

Possibly breaking changes:
 - When calling `init` the 'intermediates' collection is no longer mutable.
   Therefore, intermediates will no longer be returned from initialization by default. 
 - Don't update batch statistics during initialization.

>  - Attention: require deterministic only if using dropout

Author: Thanks for the questions, the bullet point was really unclear so I've updated it. Let me know if it is still unclear!

Other changes:
 - Rewrote various examples to use Optax instead of Flax optimizers (e.g., Imagenet, SST2)

>  - Added an NLP text classification example (SST-2 sentiment) to examples/sst2.
>    that uses a birectional LSTM (BiLSTM) to encode the input text.

Reviewer: Suggestions to improve this part:

- Fix a typo: "bidirectional" (can't keep up with all the papers, so I'm assuming you are referring to this BiLSTM: _Bidirectional Recurrent Neural Networks_ (Schuster & Paliwal, 1997) https://maxwell.ict.griffith.edu.au/spl/publications/papers/ieeesp97_schuster.pdf).
- Add a link to the SST-2 text classifier example to help users find it quicker.

```suggestion
 - Added an NLP text classification example (on the SST-2 dataset) to
   [`examples/sst2`](https://github.com/google/flax/tree/master/examples/sst2).
   that uses a bidirectional LSTM (BiLSTM) to encode the input text.
```

>  - Added flax.training.train_state to simplifying using Optax optimizers.

Reviewer: A couple of small suggestions:
- Code font for `flax.training.train_state`.
- Use "to simplify" ("...using Optax...").

```suggestion
 - Added `flax.training.train_state` to simplify using Optax optimizers.
```

 - `mutable` argument is now available on `Module.init` and `Module.init_with_outputs`

>  - Bug Fix: Correctly handle non-default parameters of Linen Modules with nested inheritance.

Reviewer: Nit:

```suggestion
 - Bug fix: Correctly handle non-default parameters of Linen Modules with nested inheritance.
```

>  - Expose dot_product_attention_weights, allowing access to attention weights.

Reviewer: Nit:

- Back ticks for `dot_product_attention_weights` for consistency.

```suggestion
 - Expose `dot_product_attention_weights`, allowing access to attention weights.
```

 - `BatchNorm` instances will behave correctly during init when called multiple times.
 - Added a more extensive "how to contribute" guide in `contributing.md`.

>  - Add proper cache behavior for lift.jit, fixing cache misses.

Reviewer: Suggestions:

- Nit (back ticks).
- Also, is it a lifted version of JIT (https://flax.readthedocs.io/en/latest/_autosummary/flax.linen.jit.html#flax.linen.jit)? Would it make sense to add a link to the API (Flax Linen), since it appears to be a non-native JIT transform? 
 
e.g.

```suggestion
 - Add proper cache behavior for [`lift.jit`](https://flax.readthedocs.io/en/latest/_autosummary/flax.linen.jit.html#flax.linen.jit),
fixing cache misses.
```

 - Fix bug in Embed layer: make sure it behaves correctly when embedding is np.array.

>  - Fix linen.Module for deep inheritance chains.

Reviewer: Nit:

```suggestion
 - Fix `linen.Module` for deep inheritance chains.
```

 - Fix bug in DenseGeneral: correctly expand bias to account for batch & noncontracting dimensions.
 - Allow Flax lifted transforms to work on partially applied Modules.

>  - Make MultiOptimizer use apply_gradient instead of apply_param_gradient

Reviewer: Nit:

- Use "code font" for `apply_gradient`, `apply_param_gradient`.
- Add a full stop.

```suggestion
 - Make MultiOptimizer use `apply_gradient` instead of `apply_param_gradient`.
```
